core/src/main/scala/org/broadinstitute/workbench/hamm/CostCalculator.scala

```diff
@@ -0,0 +1,95 @@
+import java.time.{Duration, Instant}
+import java.util.concurrent.TimeUnit
+
+import cats.data.NonEmptyList
+import cats.implicits._
+import org.broadinstitute.workbench.hamm.pricing.{ComputePriceKey, PriceList, StoragePriceKey, UsageType}
+
+import scala.concurrent.duration.FiniteDuration
+
+
+object CostCalculator {
+
+  def getPriceOfWorkflow(callMetaDataJson: MetadataResponse, priceList: PriceList): Either[Throwable, Double] = {
+    val ls: List[Either[NonEmptyList[String], Double]] = callMetaDataJson.calls.map { call =>
+      getPriceOfCall(call, priceList, callMetaDataJson.startTime, callMetaDataJson.endTime).leftMap(NonEmptyList.one)
+    }
+
+    ls.parSequence.leftMap(errors => new Exception(errors.toList.toString)).map(_.sum)
+  }
+
+  private def getPriceOfCall(call: Call, priceList: PriceList, startTime: Instant, endTime: Instant): Either[String, Double] = {
+    for {
+      _ <- if (Status.terminalStatuses.contains(call.status)) Right(()) else Left(s"Call {name} status was ${call.status.asString}.") // not evaluating calls that are in flight
+      // ToDo: handle multiple disks in one call
+      diskSize = call.runtimeAttributes.disks.diskSize.asInt + call.runtimeAttributes.bootDiskSizeGb.asInt
+      diskType = call.runtimeAttributes.disks.diskType
+      usageType = getUsageType(call)
+      computePrices <- Either.catchNonFatal(priceList.compute.computePrices.get(ComputePriceKey(call.region, call.machineType, usageType)).get).leftMap(_ => s"couldn't get compute prices for region ${call.region}, machineType ${call.machineType}, $usageType and non-extended.")
+      storagePrice <-  Either.catchNonFatal(priceList.storage.pricesByDisk.get(StoragePriceKey(call.region, diskType)).get).leftMap(_ => s"couldn't get storage prices for region ${call.region}, and diskType $diskType")
+    } yield {
+      // ToDo: calculate subworkflows
+      val wasPreempted = wasCallPreempted(call)
+      // only looking at actual and not requested disk info
+      val callDuration = getCallDuration(call, startTime, endTime)
+      // adjust the call duration to account for preemptibility - if a VM preempted less than 10 minutes after it is created, user incurs no cost
+      val adjustedCallDurationInSeconds = if (usageType == UsageType.Preemptible && wasPreempted && callDuration.toSeconds < (10 * 60)) 0 else callDuration.toSeconds
+      val cpuCost = adjustedCallDurationInSeconds * computePrices.cpu
+      val diskGbHours = call.runtimeAttributes.disks.diskSize.asInt * (adjustedCallDurationInSeconds)
+      val diskCost = diskGbHours * storagePrice
+      val memCost = adjustedCallDurationInSeconds * computePrices.ram
+      cpuCost + diskCost + memCost
+    }
+  }
+
+
+  private def wasCallPreempted(call: Call): Boolean = {
+    // treat preempted and retryableFailure as the same
+    call.executionEvents.exists(event => (event.description.asString.equals("Preempted") || event.description.asString.equals("RetryableFailure")))
```

Comment:
I would make `description` an ADT as well, and create a decoder for it

Reply:
We don't know all of the possible values for a description though